### PR TITLE
working training!

### PR DESCRIPTION
**Pull request by clee-ai**
I got the basic training running. It will train a sparseconv3d model on the s3dis1x1 dataset. Next I will get the structured configs working. I'm pretty happy with the dividing line I made between "backbone models" (the api models built by torch_points3d) and the "lightning models" (the models that define how to train the model, add appropriate classification heads, data processing, etc.). Let me know your suggestions. I also included some comments in the files you can take a look at.

training cmd:
`python train.py model=segmentation/sparseconv3d/ResUNet32 dataset=segmentation/s3dis/s3dis1x1`

> move scheduler/optim to be contained within the model, not in global scope
> 
> separate "backbones" from "lightning models"
> 
> fix sp3d sparsetensor device
> 
> add basic data transforms (need to structure better than original repo)
> 
> add training methods

### CODE REVIEW

File: torch_points3d/applications/sparseconv3d.py [hunk 1]
```diff
@@ -135,7 +135,7 @@ def _set_input(self, data):
         data:
             a dictionary that contains the data itself and its metadata information.
         """
-        self.input = sp3d.nn.SparseTensor(data.x, data.coords, data.batch)
+        self.input = sp3d.nn.SparseTensor(data.x, data.coords, data.batch, self.device)
```
Comment (humanpose1) [inline torch_points3d/applications/sparseconv3d.py:138]: I guess that SparseTensor and Pytorch-lightning interesting features are not yet totally compatible. So we need to specify in the sparseTensor the device.

File: torch_points3d/models/segmentation/base_model.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+    # Most segmentation models will share the same head and criterion
+    # Note that we don't even have to create a sparseconv3d subclass of PointCloudBaseModel
+    # because the actual training logic is contained all within this and PointCloudBaseModel
+    # so all we need to do is define a `backend` (in this case sparseconv3d)
+    def _build_model(self, backbone_cfg):
+        super()._build_model(backbone_cfg)
+
+        self.head = nn.Sequential(nn.Linear(self.backbone.output_nc, self.num_classes))
+        self.criterion = F.nll_loss
```
Comment (humanpose1) [inline torch_points3d/models/segmentation/base_model.py]: for the loss, I would also use hydra to instantiate it. For segmentation, there is just one loss, but for other tasks it has many losses

File: torch_points3d/modules/SparseConv3d/nn/torchsparse.py [hunk 3]
```diff
@@ -41,7 +41,7 @@ def __init__(
             stride=stride,
             dilation=dilation,
             bias=bias,
-            transpose=True,
+            transposed=True,
```
Comment (humanpose1) [inline torch_points3d/modules/SparseConv3d/nn/torchsparse.py:44]: which version of torchsparse do you use ?

**Review by humanpose1 — COMMENTED** [review]
Thanks for the contribution. You've added some data augmentation. But how do you plan to instantiate the transforms? Will you use hydra's instantiate function ? or the custom `instantiate_transform` ?
I'm wondering how we can perform static type checking for every transforms in the conf....

Also, it is great that you separate the backbone and the models.But I suggest to separate even more.

In your PR, you have backbone -> lightningModel
I would suggest:  backbone -> complexModel -> lightningModel
It is not a huge difference for the segmentation task. But IMHO, it would be better for tasks such as object detection or registration. For example we could have:
Pointnet++ -> votenet -> ObjectDetectionModel.
PointNet++ -> votenet2 -> objectDetectionModel
In other words, we could have the following organization. 
```
torch_points3d
 └──applications
 └──models
 └──tasks
```
application have the basic backbones unet and encoder.
models have custom complex models for different tasks, such as pointgroup or votenet (and the losses).
tasks have LightningModel to manage the optimizer, the scheduler, metrics computation....

What do you think ?

File: torch_points3d/applications/sparseconv3d.py (hunk 1 above)
Comment (clee-ai) [reply torch_points3d/applications/sparseconv3d.py:138]: It's hard to tell by the docs, but it seems like when constructing new tensors you have to specifiy the device explicitly. https://pytorch-lightning.readthedocs.io/en/latest/guides/speed.html#tranfering-tensors-to-device

File: torch_points3d/modules/SparseConv3d/nn/torchsparse.py (hunk 3 above)
Comment (clee-ai) [reply torch_points3d/modules/SparseConv3d/nn/torchsparse.py:44]: latest (1.4)

**Comment by clee-ai** [comment]
> Thanks for the contribution. You've added some data augmentation. But how do you plan to instantiate the transforms? Will you use hydra's instantiate function ? or the custom instantiate_transform ?

I think I will instantiate the transforms using hydra. Or have some custom logic in the `Instantiator` class. We can look at those in another pr.

I think adding the additional separation will work, but we need to be careful about not duplicating code within the new models. The situation I'm trying to avoid is, in the current torchpoints, each model implements similar `set_input` methods and we have a lot of inconsistency when dealing with weighted classes, ignore indexes, etc. unnecessarily. But, I think we can solve that by having strong inheritance patterns which abstract away most of that for the individual models

**Comment by clee-ai** [comment]
@humanpose1 I updated the file structure to where we have: application => model => lightning model (task), so all the model forward/init code is left seperate of ptl. Let me know how the structure looks to you. In this case, we can just use SegmentationBaseModel for sparseconv3d, as there's no special logic to the sparseconv segmentation model.

File: conf/dataset/default.yaml [hunk 4]
```diff
@@ -2,6 +2,8 @@
 # cfg:
   # torch data-loader specific arguments
 cfg: 
+  num_classes:
```
Comment (humanpose1) [inline conf/dataset/default.yaml]: `num_classes` is specific to segmentation or classification

File: conf/model/segmentation/default.yaml [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# @package model
+defaults:
+  - /model/default
+
+model:
+  _recursive_: false
+  _target_: torch_points3d.models.segmentation.base_model.SegmentationBaseModel
+  num_classes: ${dataset.cfg.num_classes}
+  criterion:
+    _target_: torch.nn.NLLLoss
```
Comment (humanpose1) [inline conf/model/segmentation/default.yaml:10]: Great !

File: torch_points3d/models/segmentation/base_model.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    def __init__(self, instantiator: Instantiator, num_classes: int, backbone: DictConfig, criterion: DictConfig):
+        super().__init__(instantiator)
+
+        print(backbone)
+        self.backbone = self.instantiator.backbone(backbone)
+        self.criterion = self.instantiator.instantiate(criterion)
+
+        self.head = nn.Sequential(nn.Linear(self.backbone.output_nc, num_classes))
+
+    def set_input(self, data):
```
Comment (humanpose1) [inline torch_points3d/models/segmentation/base_model.py]:
```python
def set_input(self, data: Data): -> None
```

File: torch_points3d/models/segmentation/base_model.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        self.batch_idx = data.batch.squeeze()
+        self.input = data
+        if data.y is not None:
+            self.labels = data.y
+        else:
+            self.labels = None
+
```
Comment (humanpose1) [inline torch_points3d/models/segmentation/base_model.py:30]:
```python
def forward(self): -> torch.Tensor
```

File: torch_points3d/models/segmentation/base_model.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
+from omegaconf import DictConfig
+
+import torch.nn as nn
+import torch.nn.functional as F
+
```
Comment (humanpose1) [inline torch_points3d/models/segmentation/base_model.py:8]:
```python
from torch_geometric.data import Data
```

File: torch_points3d/models/base_model.py [hunk 9]
```diff
@@ -1,86 +1,16 @@
...
-            num_warmup_steps *= num_training_steps
-        return num_training_steps, num_warmup_steps
+        self.instantiator = instantiator
 
-    def setup(self, stage: str):
-        self.configure_metrics(stage)
+    def set_input(self, data):
+        raise (NotImplementedError("set_input needs to be defined!"))
 
-    def configure_metrics(self, stage: str) -> Optional[Any]:
```
Comment (humanpose1) [inline torch_points3d/models/base_model.py:81]: maybe we should also have a function that return all the losses
For some models such as KPConv, there are specific [regularization losses.](https://github.com/nicolas-chaulet/torch-points3d/blob/5378a2ae7ef612433acedaa497073971ec66ff09/torch_points3d/models/segmentation/kpconv.py#L142)
. The forward will return the total loss.
```python
def get_losses(self) -> Dict[str, torch.Tensor]:
"""
"""
```
When we have multiple losses we can track them.

**Review by humanpose1 — COMMENTED** [review]
Ok great PR. It is well configured.
When you're ready you can merge

File: torch_points3d/models/base_model.py (hunk 9 above)
Comment (clee-ai) [reply torch_points3d/models/base_model.py:81]: Good idea!

**Comment by clee-ai** [comment]
Great! I will merge it now and we can update it later.
